{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "plt.ion()\n",
    "\n",
    "import logomaker as lm\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "\n",
    "msa = []\n",
    "with open(msa_file,\"r\") as f:\n",
    "    raw_seq = f.readlines()\n",
    "    for i in range(0,len(raw_seq),2):\n",
    "        seq = raw_seq[i+1]\n",
    "        ident = raw_seq[i].strip().split(\"/\")[-1]\n",
    "        msa.append(SeqRecord(Seq(seq), id=ident))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#msa_file=\"pmpnn_design/alpha_chain_top100.fasta\"\n",
    "#change to your msa file - only one chain\n",
    "msa_file=\"esm_design/alpha_chain_top100.fasta\"\n",
    "#msa_file=\"esm_design/beta_chain_top100.fasta\"\n",
    "#change to your fasta with native sequence - only one chain\n",
    "native_fasta=\"chainD.fasta\"\n",
    "\n",
    "#set range of residues, that should be displayed in the logo plot\n",
    "start = 10\n",
    "end = 20\n",
    "\n",
    "\n",
    "res_range = [i for i in range(start-1,end)]\n",
    "\n",
    "#or create list with specific residues:\n",
    "#res_range = [10,11,66,68,69,70,71,73,74,75,110,112]\n",
    "#res_range = [8,10,11,66, 69,77,88,110,112]\n",
    "\n",
    "#reads msa file and creates matrix containing counts for each mutation\n",
    "with open(msa_file,\"r\") as f:\n",
    "    raw_seq = f.readlines()\n",
    "f.close()\n",
    "\n",
    "seqs = [seq.strip() for seq in raw_seq if ('#' not in seq) and ('>') not in seq]\n",
    "seqs_clean = list(filter(None, seqs))\n",
    "counts_mat = lm.alignment_to_matrix(seqs_clean)\n",
    "counts_mat.head()\n",
    "\n",
    "#reads native sequence file and creates list of native residues to use later for axis labels \n",
    "#- native fasta should have same starting point and length as fastas from the design\n",
    "o_seq =[]\n",
    "with open(native_fasta, 'r') as fasta:\n",
    "    line = [l.strip() for l in fasta]\n",
    "    o_seq[:]= line[1]\n",
    "    #res_range=[i for i in range(1,len(o_seq)+1)]\n",
    "    o_seq_res = [o_seq[i-1] for i in res_range]\n",
    "fasta.close()\n",
    "\n",
    "#filters count matrix based on the selected residue positions\n",
    "counts_mat_range = counts_mat.loc[[x-1 for x in res_range]]\n",
    "counts_mat_range.reset_index(drop=True, inplace=True)\n",
    "\n",
    "print(f'the native residues displayed in this logo plot are {[str(i+1)+j for i,j in zip(res_range,o_seq_res)]} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#output = 'pmpnn_design_top100_alpha_logoplot.png'\n",
    "output = \"esm_design_top100_alpha_completelogoplot\" #change to output name\n",
    "\n",
    "cm = 1/2.54\n",
    "fig, ax = plt.subplots(figsize=[10*cm,5*cm])\n",
    "seq_logo = lm.Logo(counts_mat_range,\n",
    "                           ax=ax,\n",
    "                           stack_order='small_on_top',\n",
    "                          #fade_probabilities=True,\n",
    "                           #font_name='Times New Roman',\n",
    "                           color_scheme='black',\n",
    "                           vpad=.02,\n",
    "                           #color_scheme='grays'\n",
    "                          )\n",
    "seq_logo.ax.set_yticks([0, 1])\n",
    "seq_logo.ax.set_yticklabels(['0', '1'], size=16)\n",
    "seq_logo.ax.set_xticks(range(len(res_range)))\n",
    "seq_logo.ax.set_xticklabels(o_seq_res, size=16)\n",
    "secax = ax.secondary_xaxis('top')\n",
    "secax.set_xticks(range(len(res_range)))\n",
    "secax.set_xticklabels(res_range, size=16)\n",
    "secax.set_xlabel(\"Residue numbers\", size=18, labelpad=10)\n",
    "ax.set_xlabel(\"Wildtype amino acid\", size=18, labelpad=10)\n",
    "ax.set_ylabel(\"count\", size=18, labelpad=10)\n",
    "ax.tick_params(axis=u'both', which=u'both',length=0)\n",
    "secax.tick_params(axis=u'both', which=u'both',length=0)\n",
    "\n",
    "\n",
    "#select which residues to highlight\n",
    "highlights = []\n",
    "nothighlights = [i for i in range(len(res_tange)) if i not in highlights]\n",
    "for i in highlights:\n",
    "    seq_logo.highlight_position(p=i-1, color='#9932CC', alpha=.5)\n",
    "for i in nothighlights:\n",
    "    seq_logo.highlight_position(p=i-1, color='#FFFFFF', alpha=.5)\n",
    "\n",
    "    \n",
    "fig.tight_layout()\n",
    "fig.savefig(f'{output}.png', dpi=300) #save as png\n",
    "fig.savefig(f'{output}.svg', dpi=300) #save as vector graphic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
